File: src/repository/BusTerminalDao.java
package repository;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.function.Consumer;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.EntityTransaction;
import javax.persistence.Persistence;
import javax.persistence.Query;

import model.BusTerminal;
import model.User;

public class BusTerminalDao implements Repository<BusTerminal, Integer>{

	private EntityManagerFactory emf = Persistence.createEntityManagerFactory(
			"jdbc:sqlserver://localhost/Alibaba.odb;integratedSecurity=true;");
	private EntityManager em=emf.createEntityManager();  
	
	@Override
	public BusTerminal findbyId(Integer id) {
		try {
			return em.find(BusTerminal.class, id);
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public List<BusTerminal> findByIDs(Collection<Integer> ids) {
		List<BusTerminal> busTerminals = new ArrayList<BusTerminal>();
		try {
			for(int id : ids) {
				busTerminals.add(em.find(BusTerminal.class, id));
			}
			return busTerminals;
		} catch (Exception e) {
			return null;
		}
	}

	@Override
	public List<BusTerminal> findAll() {
		Query q = em.createQuery("select * from Bus_Terminal");
		return q.getResultList();
	}

	@Override
	public Boolean deleteByID(Integer id) {
		try {
			executeInsideTransaction(entityManager -> em.remove(em.find(BusTerminal.class, id)));
			return true;
		} catch(Exception e) {
			return false;
		}
	}

	@Override
	public Boolean DeleteByIDs(Collection<Integer> ids) {
		try {
			for(int id : ids) {
				executeInsideTransaction(entityManager -> em.remove(em.find(BusTerminal.class, id)));
			}
			return true;
		} catch(Exception e) {
			return false;
		}
	}

	@Override
	public BusTerminal save(BusTerminal E) {
		try {
			executeInsideTransaction(entityManager -> entityManager.persist(E));
			return em.find(BusTerminal.class, emf.getPersistenceUnitUtil().getIdentifier(E));
		} catch(Exception e) {
			executeInsideTransaction(entityManager -> entityManager.merge(E));
			return em.find(BusTerminal.class, emf.getPersistenceUnitUtil().getIdentifier(E));
		}
	}
	
	private void executeInsideTransaction(Consumer<EntityManager> action) {
        EntityTransaction tx = em.getTransaction();
        try {
            tx.begin();
            action.accept(em);
            tx.commit(); 
        }
        catch (RuntimeException e) {
            tx.rollback();
            throw e;
        }
    }

}

File: repository/BusTerminalDao.py
from Repository import RepositoryInterface
from Connection import getConn
from BusTerminal import BusTerminal
class BusTerminalDao(RepositoryInterface):
    def findById(id):
        try:
            c = getConn()
            c.execute('select * from Bus_Terminal where id=' + str(id))
            i = c.fetchall()
            return BusTerminal(i[0][0], i[0][1], i[0][2])
        except:
            print("There was an error")
            return None
    def findByIds(ids):
        try:
            a = []
            c = getConn()
            for id in ids:
                c.execute('select * from Bus_Terminal where id=' + str(id))
                i = c.fetchall()
                a.append(BusTerminal(i[0][0], i[0][1], i[0][2]))
            return a
        except:
            print("There was an error")
            return None
    def findAll():
        try:
            a = []
            c=getConn()
            c.execute('select * from Bus_Terminal')    
            for i in c:
                a.append(BusTerminal(i[0], i[1], i[2]))
            return a
        except:
            print("There was an error")
            return None
    def deleteById(id):
        try:
            c=getConn()
            c.execute('delete from Bus_Terminal where id=' + str(id))
            return True
        except:
            return False
    def deleteByIds(ids):
        try:
            c=getConn()
            for id in ids:
                   c.execute('delete from Bus_Terminal where id=' + str(id))
            return True
        except:
            return False
    def save(entity):
        try:
            c=getConn()
            c.execute('select * from Bus_Terminal where id=' + str(entity.id))
            i = c.fetchall()
            if len(i) == 0:
                c.execute('insert into Bus_Terminal values(' + str(entity.name)+','+ str(entity.cityId)+')')
            else :
                c.execute('update Bus_Terminal set name=' +str(entity.name)+', cityId='+ str(entity.cityId)+' where id='+ str(entity.id))
            return BusTerminal(entity.id, entity.name, entity.cityId)
        except:
            print("There was an error")
            return None